Implement watch function

bodograumann opened the pull request:
So this is a basic implementation which works for me and does not give any type errors so far :-D
It’s missing documentation and tests, which I can do as soon as the implementation is finished.

I’m not completely happy with parts of this yet though.
- Defining an empty prototype in `VuexModule` seems weird.
- Also the types in the actual implementation are not perfect, as I was not able use generics there. Maybe it makes sense to use `function` instead of `=>` at that point; or could that pose a problem?

Concerns #4

bodograumann committed:
Implement watch api

gertqin commented on src/module-factory.ts:
```diff
@@ -235,6 +239,19 @@ export class VuexClassModuleFactory {
+        options?: WatchOptions
+      ) => {
+        watch(
+          () => fn(obj), callback, options
```
Although it probably doesn't make a difference, I would prefer for consistency that the proxy obj uses the function parameters provided by vuex, instead of just the 'global' proxy. Something like
```ts
(state: any, getters: any) => fn(this.buildThisProxy({ state, getters })), callback, options
```
similar to how the getters proxy work.

bodograumann replied:
Yes, I had the same thought :-D

gertqin commented on src/module-factory.ts:
```diff
@@ -181,6 +184,7 @@ export class VuexClassModuleFactory {
     const accessorModule = this.buildThisProxy({
       ...store,
       state: store.state[name],
+      watch: store.watch,
```
I'd prefer to have the definition of ```$watch``` to be in the ```buildAccessor``` function, as it does not have anything to do with the other ```this``` proxy objects. Something like
```ts
accessorModule.$watch = ...
```
and then remove the ```watch``` field of ```StoreProxyDefinition```

bodograumann replied:
Ok, I will change that.

gertqin reviewed (changes requested):
Nice work! :) 
In general it looks good, but I'd like a few changes, pointed out by my comments.
(This is my first pull request review - so hopefully I've done it appropriately)

gertqin commented:
And
* I'm fine with the empty function providing the correct typings in the VuexModule,
* It's fine that the typings in the actual implementation are not perfect, as long as it works :) It's only the exposed typings that should be as good as possible.

gertqin commented:
Also are you using VS Code and if so do you have the Prettier extension installed? It looks like your formatting is a bit different from mine.

If you haven't its fine, I can change the formatting myself after the pull request is merged.

bodograumann commented:
I’m using vim, not VS Code. For my own code I am also using prettier, but through eslint, because that is what `vue create` supplies out of the box.
It seems you have a configuration for tslint, but running `npm run lint` did not do much. I might have a look at that later.

bodograumann commented:
It seems in contrast to the eslint setup provided by vue-cli, this tslint does not run prettier. For that you would need [tslint-plugin-prettier](https://github.com/prettier/tslint-plugin-prettier).

I ran prettier and it turns out that this also changes the formating of existing code. Do you maybe have a custom prettier config? I'll push the result, when I made the requested changes.

bodograumann committed:
Add prettier tslint plugin

gertqin commented:
I haven't spent time on setting up prettier & linting so that it should work in any other environment than vs code (as I just use the same setup for all my projects), so for now I don't think you should spend time on formatting.

I'll try to find time to fix ```npm run lint```.

EDIT: But if you are already looking at it, the only prettier rule I have is ```printWidth=120```. The rest is just the tslint rules provided in ```tslint.json```.

bodograumann committed:
Use vuex version with #1058

There is an unreleased fix to have the getters argument on the watch
callback. See https://github.com/vuejs/vuex/pull/1058

bodograumann commented:
It's probably the `printwidth` then. I can add that I think.

Unfortunately the version which uses the state and getters from the callback directly doesn't work yet. I have to debug that.

gertqin commented:
Might be because you need ```useNamespaceKey: true```?

bodograumann commented:
Ah yes, that was it. Thanks for the hint, I completely missed the condition there :-D

bodograumann committed:
Add test for watch api

bodograumann committed:
Add documentation for the watch api

bodograumann commented:
Ok, linting, test and docs are there. Please have another look.

gertqin commented:
Looks very good! Nice work, really appreciated.

bodograumann commented:
Cool :-) Would you mind publishing it to npm?

gertqin commented:
Working on it :)
